### PR TITLE
Add AMP inference to MonoT5, DuoT5 and MonoBERT

### PR DESCRIPTION
**Pull request by hugoabonizio**
Following #200, this PR adds AMP inference to MonoT5, DuoT5 and MonoBERT (disabled by default to avoid a performance drop).

### CODE REVIEW

**Comment by ronakice** [comment]
Hey @hugoabonizio ! Thanks a lot fo this feature again but can you give me some numbers, runtime with/without, performance with/without for three models?

File: pygaggle/rerank/transformer.py [hunk 1]
```diff
@@ -34,10 +34,12 @@
 class MonoT5(Reranker):
     def __init__(self,
                  model: T5ForConditionalGeneration = None,
-                 tokenizer: QueryDocumentBatchTokenizer = None):
+                 tokenizer: QueryDocumentBatchTokenizer = None,
+                 use_amp=False):
```
Comment (rodrigonogueira4) [inline pygaggle/rerank/transformer.py:38]:
```suggestion
                 use_amp = False):
```

File: pygaggle/rerank/transformer.py [hunk 2]
```diff
@@ -80,10 +83,12 @@ def rescore(self, query: Query, texts: List[Text]) -> List[Text]:
 class DuoT5(Reranker):
     def __init__(self,
                  model: T5ForConditionalGeneration = None,
-                 tokenizer: QueryDocumentBatchTokenizer = None):
+                 tokenizer: QueryDocumentBatchTokenizer = None,
+                 use_amp=False):
```
Comment (rodrigonogueira4) [inline pygaggle/rerank/transformer.py:87]:
```suggestion
                 use_amp = False):
```

File: pygaggle/rerank/transformer.py [hunk 3]
```diff
@@ -179,10 +185,12 @@ def rescore(self, query: Query, texts: List[Text]) -> List[Text]:
 class MonoBERT(Reranker):
     def __init__(self,
                  model: PreTrainedModel = None,
-                 tokenizer: PreTrainedTokenizer = None):
+                 tokenizer: PreTrainedTokenizer = None,
+                 use_amp=False):
```
Comment (rodrigonogueira4) [inline pygaggle/rerank/transformer.py:189]:
```suggestion
                 use_amp = False):
```

**Review by rodrigonogueira4 — APPROVED** [review]
Awesome, thanks for adding this. Just a minor comment.

**Comment by hugoabonizio** [comment]
Hi @ronakice! I'm comparing MonoT5 with and without AMP enabled on a V100. I'm currently not able to compare MonoBERT and DuoT5 in our setup because they are much slower than MonoT5 and became impractible (so there's room for optimizing the implementation, e.g., adding batch predictions to MonoBERT). Since I don't have numbers to compare MonoBERT and DuoT5 I can remove this AMP feature for them, WDYT?

- **MonoT5**
  - FP32: 7.3s (±1.1s)
  - AMP: 6.0s (±1.2s)

The performance is about 99.8% of the original in our private dataset.

**Comment by ronakice** [comment]
It's fine we can leave them there but document it later, I think the speedup is interesting, thanks for running this! I'll merge!

**Pull request merged**